<a href="https://colab.research.google.com/github/darioLabrador/computer_vision_practice/blob/main/CSC3831_Practical_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# CSC3831 - Practical 2
Hello Class, welcome to the second practical session. For this practical please read the information in the text blocks, run the code blocks in sequence, and attempt the optional challenges if you have time.

In this part of the course we will use a few Python packages. The packages used are as follows:
- [Pandas](https://pandas.pydata.org/): Data manipulation and analysis tool
- [RecordLinkage](https://recordlinkage.readthedocs.io/en/latest/): A python Record Linkage toolset
- [FuzzyMatcher](https://github.com/RobinL/fuzzymatcher): Legacy record linkage tool, no longer supported
- [MissingNo](https://github.com/ResidentMario/missingno): Graphical tool for easily visualising missing data
- [Numpy](https://numpy.org/): Numerical computing tool
-

Additonally we will be playing around with some more Pandas DataFrame manipulations. Not every method is explained in detail, so if you are interested or want to know more please look at the Pandas method documentation.

The data we will be using today are `hospital_accounts` which contains account information for various hospitals in the USA, and `hospital_reimbursement` which contains reimbursement financial data for various hospitals in the USA.


# Record Linkage

Firstly let's install the matching tools we'll use in today's practical.

In [1]:
!pip install recordlinkage

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 926.9/926.9 kB 17.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 355.9/355.9 kB 20.8 MB/s eta 0:00:00


In [2]:
!pip install fuzzymatcher

  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 159.9/159.9 kB 5.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.2/3.2 MB 52.3 MB/s eta 0:00:00
  Created wheel for metaphone: filename=Metaphone-0.6-py3-none-any.whl size=13901 sha256=4bea7a1b505428f45d2d44cc6fdaa63fd955ba65bd516ee60d1fe7e7a22dc637
  Stored in directory: /root/.cache/pip/wheels/d0/fc/72/0c9e56e8c85dc1b2998f4c8f5e9075b8eb71b046b519aa1759
Successfully built metaphone


Now let's load the standard Pandas package and the datasets we're looking to join together.

We want to combine together the two datasets, but they don't have numeric indices that correspond. So how do we go about doing this?

In [3]:
import pandas as pd

In [4]:
hospital_accounts = pd.read_csv('https://github.com/chris1610/pbpython/raw/master/data/hospital_account_info.csv')
hospital_reimbursement = pd.read_csv('https://raw.githubusercontent.com/chris1610/pbpython/master/data/hospital_reimbursement.csv')

In [5]:
hospital_accounts.head()

Account_Num                      Facility Name                    Address  \
0        10605             SAGE MEMORIAL HOSPITAL  STATE ROUTE 264 SOUTH 191   
1        24250        WOODRIDGE BEHAVIORAL CENTER       600 NORTH 7TH STREET   
2        10341           DOUGLAS GARDENS HOSPITAL            5200 NE 2ND AVE   
3        81101  SUNCOAST BEHAVIORAL HEALTH CENTER             4480 51ST ST W   
4        39835           TREASURE VALLEY HOSPITAL   8800 WEST EMERALD STREET   

           City State  ZIP Code County Name    Phone Number  \
0        GANADO    AZ     86505      APACHE  (928) 755-4541   
1  WEST MEMPHIS    AR     72301  CRITTENDEN  (870) 394-4113   
2         MIAMI    FL     33137  MIAMI-DADE  (305) 751-8626   
3     BRADENTON    FL     34210     MANATEE  (941) 792-2222   
4         BOISE    ID     83704         ADA  (208) 373-5000   

               Hospital Type              Hospital Ownership  
0  Critical Access Hospitals  Voluntary non-profit - Private  
1                Psychiatric                     Proprietary  
2       Acute Care Hospitals  Voluntary non-profit - Private  
3                Psychiatric                     Proprietary  
4       Acute Care Hospitals                     Proprietary

In [6]:
hospital_reimbursement.head()

Provider_Num                     Provider Name     Provider Street Address  \
0        839987  SOUTHEAST ALABAMA MEDICAL CENTER      1108 ROSS CLARK CIRCLE   
1        519118     MARSHALL MEDICAL CENTER SOUTH  2505 U S HIGHWAY 431 NORTH   
2        733073    ELIZA COFFEE MEMORIAL HOSPITAL          205 MARENGO STREET   
3        201752          MIZELL MEMORIAL HOSPITAL               702 N MAIN ST   
4        678488                 ST VINCENT'S EAST  50 MEDICAL PARK EAST DRIVE   

  Provider City Provider State  Provider Zip Code  Total Discharges  \
0        DOTHAN             AL              36301               118   
1          BOAZ             AL              35957                43   
2      FLORENCE             AL              35631                73   
3           OPP             AL              36467                12   
4    BIRMINGHAM             AL              35235                74   

   Average Covered Charges  Average Total Payments  Average Medicare Payments  
0                 20855.61                 5026.19                    4115.52  
1                 13289.09                 5413.63                    4490.93  
2                 22261.60                 4922.18                    4021.79  
3                 10901.33                 5343.50                    4284.17  
4                 28117.95                 5947.12                    4819.53

The "Account_Num" and "Provider_Num" are indices to be used within the files themselves, but don't nicely match up across files, making this a Record Linkage problem. It looks like there are two name columns, "Facility Name" and "Provider Name" with which we should be able to identify if two records represent the same real-world entity.

Both files use the same spacing and capitilisation format throughout each column so we don't need to do any pre-processing.

Lets see how many records exist in each dataset.

In [7]:
print("raw dataframe number of records:")
print("hospital_accounts:  {x}".format(x=hospital_accounts['Account_Num'].count()))
print("hospital_reimbursement:  {x}".format(x=hospital_reimbursement['Provider_Num'].count()))

raw dataframe number of records:
hospital_accounts:  5339
hospital_reimbursement:  2697


It looks like the `hospital_reimbursements` DataFrame is has less records, so we can try joining based on its keys.

First we just explore the data and then try exact joins.
Focusing on "Facility Name" and "Provider Name" we'll just pick providers starting with "AD" from both (to look at a few examples), and pick the top 20 to visually compare.

In [8]:
cond_accounts = hospital_accounts['Facility Name'].str.startswith("AD")
hospital_accounts.loc[cond_accounts].sort_values(by=['Facility Name'], ascending=True)[['Facility Name', 'Address', 'City']].head(20)

Facility Name  \
3282                          AD HOSPITAL EAST, LLC   
736   ADAIR ACUTE CARE AT OSAWATOMIE STATE HOSPITAL   
2515                 ADAIR COUNTY MEMORIAL HOSPITAL   
3915           ADAMS COUNTY REGIONAL MEDICAL CENTER   
280                         ADAMS MEMORIAL HOSPITAL   
3017               ADCARE HOSPITAL OF WORCESTER INC   
2883                      ADENA PIKE MEDICAL CENTER   
4211                  ADENA REGIONAL MEDICAL CENTER   
4887       ADIRONDACK MEDICAL CENTER - SARANAC LAKE   
4429         ADMIN DE SERVICIOS MEDICOS  PUERTO RIC   
593                      ADVANCED SURGICAL HOSPITAL   
4147                       ADVENTHEALTH CARROLLWOOD   
71                       ADVENTHEALTH CENTRAL TEXAS   
1362                         ADVENTHEALTH DADE CITY   
2751                     ADVENTHEALTH DAYTONA BEACH   
5035                            ADVENTHEALTH DELAND   
4963                            ADVENTHEALTH DURAND   
1692                     ADVENTHEALTH FISH MEMORIAL   
2894                            ADVENTHEALTH GORDON   
4536                  ADVENTHEALTH HEART OF FLORIDA   

                              Address           City  
3282    12950 EAST FREEWAY, SUITE 100        HOUSTON  
736          500 STATE HOSPITAL DRIVE     OSAWATOMIE  
2515                      609 SE KENT     GREENFIELD  
3915         230 MEDICAL CENTER DRIVE         SEAMAN  
280                   1100 MERCER AVE        DECATUR  
3017               107 LINCOLN STREET      WORCESTER  
2883                    100 DAWN LANE        WAVERLY  
4211                272 HOSPITAL ROAD    CHILLICOTHE  
4887  2233 STATE ROUTE 86, PO BOX 471   SARANAC LAKE  
4429         BO MONACILLO CARR NUM 22       SAN JUAN  
593                   100 TRICH DRIVE     WASHINGTON  
4147            7171 N DALE MABRY HWY          TAMPA  
71            2201 S CLEAR CREEK ROAD        KILLEEN  
1362                 13100 FT KING RD      DADE CITY  
2751     301 MEMORIAL MEDICAL PARKWAY  DAYTONA BEACH  
5035               701 W PLYMOUTH AVE         DELAND  
4963                   1220 3RD AVE W         DURAND  
1692                  1055 SAXON BLVD    ORANGE CITY  
2894                1035 RED BUD ROAD        CALHOUN  
4536                     40100 HWY 27      DAVENPORT

In [9]:
cond_reimbursement = hospital_reimbursement['Provider Name'].str.startswith('AD')
hospital_reimbursement.loc[cond_reimbursement].sort_values(by=['Provider Name'], ascending=True)[['Provider Name','Provider Street Address', 'Provider City']].head(20)

Provider Name  \
1883                      ADENA REGIONAL MEDICAL CENTER   
1613                          ADIRONDACK MEDICAL CENTER   
841                      ADVENTIST BOLINGBROOK HOSPITAL   
840                                  ADVENTIST GLENOAKS   
1150    ADVENTIST HEALTHCARE SHADY GROVE MEDICAL CENTER   
1124  ADVENTIST HEALTHCARE WASHINGTON ADVENTIST HOSP...   
774                         ADVENTIST HINSDALE HOSPITAL   
752               ADVENTIST LA GRANGE MEMORIAL HOSPITAL   
204                            ADVENTIST MEDICAL CENTER   
1985                           ADVENTIST MEDICAL CENTER   
777                     ADVOCATE BROMENN MEDICAL CENTER   
812           ADVOCATE CHRIST HOSPITAL & MEDICAL CENTER   
810                     ADVOCATE CONDELL MEDICAL CENTER   
836                    ADVOCATE GOOD SAMARITAN HOSPITAL   
839                     ADVOCATE GOOD SHEPHERD HOSPITAL   
802            ADVOCATE ILLINOIS MASONIC MEDICAL CENTER   
818                  ADVOCATE LUTHERAN GENERAL HOSPITAL   
827                    ADVOCATE SOUTH SUBURBAN HOSPITAL   
742                           ADVOCATE TRINITY HOSPITAL   

         Provider Street Address     Provider City  
1883           272 HOSPITAL ROAD       CHILLICOTHE  
1613         2233 STATE ROUTE 86      SARANAC LAKE  
841      500 REMINGTON BOULEVARD       BOLINGBROOK  
840          701 WINTHROP AVENUE  GLENDALE HEIGHTS  
1150   9901 MEDICAL CENTER DRIVE         ROCKVILLE  
1124         7600 CARROLL AVENUE       TAKOMA PARK  
774             120 NORTH OAK ST          HINSDALE  
752     5101 S WILLOW SPRINGS RD         LA GRANGE  
204               115 MALL DRIVE           HANFORD  
1985      10123 SE MARKET STREET          PORTLAND  
777         1304 FRANKLIN AVENUE            NORMAL  
812           4440 W 95TH STREET          OAK LAWN  
810          801 S MILWAUKEE AVE      LIBERTYVILLE  
836         3815 HIGHLAND AVENUE     DOWNERS GROVE  
839          450 WEST HIGHWAY 22        BARRINGTON  
802   836 WEST WELLINGTON AVENUE           CHICAGO  
818             1775 DEMPSTER ST        PARK RIDGE  
827           17800 S KEDZIE AVE       HAZEL CREST  
742               2320 E 93RD ST           CHICAGO

We can see there's some overlap in "Facility Name" and "Provider Name", so a join should be possible.

To begin with, lets try an outer join on the "Facility Name" And "Provider Name" columns.

The [merge](https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html) method can join two DataFrame objects together using the normal SQL joins. We specifcy two columns from each DataFrame to join on (i.e. to match) and which DataFrame's keys to use.

In [10]:
merged = hospital_accounts.merge(hospital_reimbursement, left_on = 'Facility Name', right_on='Provider Name',  how='right')
merged.head(20)

Account_Num                     Facility Name  \
0       46115.0  SOUTHEAST ALABAMA MEDICAL CENTER   
1           NaN                               NaN   
2           NaN                               NaN   
3       51563.0          MIZELL MEMORIAL HOSPITAL   
4       56456.0                 ST VINCENT'S EAST   
5       88941.0    DEKALB REGIONAL MEDICAL CENTER   
6       16452.0     SHELBY BAPTIST MEDICAL CENTER   
7       29516.0    HELEN KELLER MEMORIAL HOSPITAL   
8       65285.0               DALE MEDICAL CENTER   
9       45584.0      BAPTIST MEDICAL CENTER SOUTH   
10      25491.0     JACKSON HOSPITAL & CLINIC INC   
11      16052.0       EAST ALABAMA MEDICAL CENTER   
12      70073.0    UNIVERSITY OF ALABAMA HOSPITAL   
13      93383.0            COMMUNITY HOSPITAL INC   
14      13872.0   CULLMAN REGIONAL MEDICAL CENTER   
15          NaN                               NaN   
16      90353.0    STRINGFELLOW MEMORIAL HOSPITAL   
17      22283.0               HUNTSVILLE HOSPITAL   
18      59503.0   GADSDEN REGIONAL MEDICAL CENTER   
19      20607.0    MARION REGIONAL MEDICAL CENTER   

                         Address        City State  ZIP Code County Name  \
0         1108 ROSS CLARK CIRCLE      DOTHAN    AL   36301.0     HOUSTON   
1                            NaN         NaN   NaN       NaN         NaN   
2                            NaN         NaN   NaN       NaN         NaN   
3                  702 N MAIN ST         OPP    AL   36467.0   COVINGTON   
4     50 MEDICAL PARK EAST DRIVE  BIRMINGHAM    AL   35235.0   JEFFERSON   
5           200 MED CENTER DRIVE  FORT PAYNE    AL   35968.0     DE KALB   
6        1000 FIRST STREET NORTH   ALABASTER    AL   35007.0      SHELBY   
7   1300 SOUTH MONTGOMERY AVENUE   SHEFFIELD    AL   35660.0     COLBERT   
8               126 HOSPITAL AVE       OZARK    AL   36360.0        DALE   
9      2105 EAST SOUTH BOULEVARD  MONTGOMERY    AL   36116.0  MONTGOMERY   
10              1725 PINE STREET  MONTGOMERY    AL   36106.0  MONTGOMERY   
11        2000 PEPPERELL PARKWAY     OPELIKA    AL   36801.0         LEE   
12         619 SOUTH 19TH STREET  BIRMINGHAM    AL   35233.0   JEFFERSON   
13           805 FRIENDSHIP ROAD   TALLASSEE    AL   36078.0      ELMORE   
14      1912 ALABAMA HIGHWAY 157     CULLMAN    AL   35058.0     CULLMAN   
15                           NaN         NaN   NaN       NaN         NaN   
16              301 EAST 18TH ST    ANNISTON    AL   36201.0     CALHOUN   
17                 101 SIVLEY RD  HUNTSVILLE    AL   35801.0     MADISON   
18          1007 GOODYEAR AVENUE     GADSDEN    AL   35903.0      ETOWAH   
19    1256 MILITARY STREET SOUTH    HAMILTON    AL   35570.0      MARION   

      Phone Number         Hospital Type  \
0   (334) 793-8701  Acute Care Hospitals   
1              NaN                   NaN   
2              NaN                   NaN   
3   (334) 493-3541  Acute Care Hospitals   
4   (205) 838-3122  Acute Care Hospitals   
5   (256) 845-3150  Acute Care Hospitals   
6   (205) 620-8100  Acute Care Hospitals   
7   (256) 386-4556  Acute Care Hospitals   
8   (334) 774-2601  Acute Care Hospitals   
9   (334) 288-2100  Acute Care Hospitals   
10  (334) 293-8000  Acute Care Hospitals   
11  (334) 749-3411  Acute Care Hospitals   
12  (205) 934-4011  Acute Care Hospitals   
13  (334) 283-6541  Acute Care Hospitals   
14  (256) 737-2000  Acute Care Hospitals   
15             NaN                   NaN   
16  (256) 235-8900  Acute Care Hospitals   
17  (256) 265-1000  Acute Care Hospitals   
18  (256) 494-4000  Acute Care Hospitals   
19  (205) 921-6200  Acute Care Hospitals   

                             Hospital Ownership  Provider_Num  \
0   Government - Hospital District or Authority        839987   
1                                           NaN        519118   
2                                           NaN        733073   
3                Voluntary non-profit - Private        201752   
4                Voluntary non-profit - Pri

Just looking at the top 20 records, we can see that rows 1, 2, and 15 have NaN values in several key columns. This implies there are records that exist in the right DataFrame (`hospital_reimbursements`) which don't appear in the left (`hospital_accounts`). If that is true, then there should be more columsn that are exclusive to the right DataFrame in the merged result.


In [11]:
merged.count()

Account_Num                  2739
Facility Name                2739
Address                      2739
City                         2739
State                        2739
ZIP Code                     2739
County Name                  2739
Phone Number                 2739
Hospital Type                2739
Hospital Ownership           2739
Provider_Num                 3140
Provider Name                3140
Provider Street Address      3140
Provider City                3140
Provider State               3140
Provider Zip Code            3140
Total Discharges             3140
Average Covered Charges      3140
Average Total Payments       3140
Average Medicare Payments    3140
dtype: int64

Removing these incomplete rows allows us to curate a complete dataset for further analysis, and as in lieu of contacting the persons responsible for collecting the `hospital_accounts` data and asking for the missing hospitals there's no way to impute, it is appropriate.

In [12]:
merged.dropna(inplace=True)
merged.reset_index(inplace=True)
# Using the .reset_index function moves the old indices to a column called 'index'.
# Repeating this cell multiple times causes it to crash as there's already an 'index' column, so we'll drop it to prevent that.
merged.drop(['index'], axis=1, inplace=True)
merged.head(20)

Account_Num                      Facility Name  \
0       46115.0   SOUTHEAST ALABAMA MEDICAL CENTER   
1       51563.0           MIZELL MEMORIAL HOSPITAL   
2       56456.0                  ST VINCENT'S EAST   
3       88941.0     DEKALB REGIONAL MEDICAL CENTER   
4       16452.0      SHELBY BAPTIST MEDICAL CENTER   
5       29516.0     HELEN KELLER MEMORIAL HOSPITAL   
6       65285.0                DALE MEDICAL CENTER   
7       45584.0       BAPTIST MEDICAL CENTER SOUTH   
8       25491.0      JACKSON HOSPITAL & CLINIC INC   
9       16052.0        EAST ALABAMA MEDICAL CENTER   
10      70073.0     UNIVERSITY OF ALABAMA HOSPITAL   
11      93383.0             COMMUNITY HOSPITAL INC   
12      13872.0    CULLMAN REGIONAL MEDICAL CENTER   
13      90353.0     STRINGFELLOW MEMORIAL HOSPITAL   
14      22283.0                HUNTSVILLE HOSPITAL   
15      59503.0    GADSDEN REGIONAL MEDICAL CENTER   
16      20607.0     MARION REGIONAL MEDICAL CENTER   
17      81775.0             FAYETTE MEDICAL CENTER   
18      70746.0  RIVERVIEW REGIONAL MEDICAL CENTER   
19      77492.0  RIVERVIEW REGIONAL MEDICAL CENTER   

                         Address        City State  ZIP Code County Name  \
0         1108 ROSS CLARK CIRCLE      DOTHAN    AL   36301.0     HOUSTON   
1                  702 N MAIN ST         OPP    AL   36467.0   COVINGTON   
2     50 MEDICAL PARK EAST DRIVE  BIRMINGHAM    AL   35235.0   JEFFERSON   
3           200 MED CENTER DRIVE  FORT PAYNE    AL   35968.0     DE KALB   
4        1000 FIRST STREET NORTH   ALABASTER    AL   35007.0      SHELBY   
5   1300 SOUTH MONTGOMERY AVENUE   SHEFFIELD    AL   35660.0     COLBERT   
6               126 HOSPITAL AVE       OZARK    AL   36360.0        DALE   
7      2105 EAST SOUTH BOULEVARD  MONTGOMERY    AL   36116.0  MONTGOMERY   
8               1725 PINE STREET  MONTGOMERY    AL   36106.0  MONTGOMERY   
9         2000 PEPPERELL PARKWAY     OPELIKA    AL   36801.0         LEE   
10         619 SOUTH 19TH STREET  BIRMINGHAM    AL   35233.0   JEFFERSON   
11           805 FRIENDSHIP ROAD   TALLASSEE    AL   36078.0      ELMORE   
12      1912 ALABAMA HIGHWAY 157     CULLMAN    AL   35058.0     CULLMAN   
13              301 EAST 18TH ST    ANNISTON    AL   36201.0     CALHOUN   
14                 101 SIVLEY RD  HUNTSVILLE    AL   35801.0     MADISON   
15          1007 GOODYEAR AVENUE     GADSDEN    AL   35903.0      ETOWAH   
16    1256 MILITARY STREET SOUTH    HAMILTON    AL   35570.0      MARION   
17      1653 TEMPLE AVENUE NORTH     FAYETTE    AL   35555.0     FAYETTE   
18        600 SOUTH THIRD STREET     GADSDEN    AL   35901.0      ETOWAH   
19            158 HOSPITAL DRIVE    CARTHAGE    TN   37030.0       SMITH   

      Phone Number              Hospital Type  \
0   (334) 793-8701       Acute Care Hospitals   
1   (334) 493-3541       Acute Care Hospitals   
2   (205) 838-3122       Acute Care Hospitals   
3   (256) 845-3150       Acute Care Hospitals   
4   (205) 620-8100       Acute Care Hospitals   
5   (256) 386-4556       Acute Care Hospitals   
6   (334) 774-2601       Acute Care Hospitals   
7   (334) 288-2100       Acute Care Hospitals   
8   (334) 293-8000       Acute Care Hospitals   
9   (334) 749-3411       Acute Care Hospitals   
10  (205) 934-4011       Acute Care Hospitals   
11  (334) 283-6541       Acute Care Hospitals   
12  (256) 737-2000       Acute Care Hospitals   
13  (256) 235-8900       Acute Care Hospitals   
14  (256) 265-1000       Acute Care Hospitals   
15  (256) 494-4000       Acute Care Hospitals   
16  (205) 921-6200       Acute Care Hospitals   
17  (205) 932-5966       Acute Care Hospitals   
18  (256) 543-5200       Acute Care Hospitals   
19  (615) 735-9815  Critical Access Hospitals   

                             Hospital Ownership  Provider_Num  \
0   Government - Hospital District or Authority        839987   
1                Voluntary non-profit - Private        201752   
2                Voluntary non-profit - Private

Functionally this is no different from performing an outter left join and dropping columns, as the unique rows to the `hospital_accounts` data would be removed as well.

**Optional Challenge**: Try replicating the same thing using outter left joins in the code snippet below.
- If the merged DataFrame is sorted differently, try looking at the Pandas Merge method documentation to figure out why

In [16]:
right_join = hospital_accounts.merge(hospital_reimbursement, left_on = 'Facility Name', right_on='Provider Name',  how='left')
right_join.head(20)

Account_Num                           Facility Name  \
0         10605                  SAGE MEMORIAL HOSPITAL   
1         24250             WOODRIDGE BEHAVIORAL CENTER   
2         10341                DOUGLAS GARDENS HOSPITAL   
3         81101       SUNCOAST BEHAVIORAL HEALTH CENTER   
4         39835                TREASURE VALLEY HOSPITAL   
5         48460                  INTERMOUNTAIN HOSPITAL   
6         30098  CHOATE MENTAL HEALTH & DEVELOPMENT CTR   
7         92639        GOODLAND REGIONAL MEDICAL CENTER   
8         86448                  SPRING HARBOR HOSPITAL   
9         28455           EASTERN SHORE HOSPITAL CENTER   
10        79639         CHOCTAW REGIONAL MEDICAL CENTER   
11        58893             ST LOUIS CHILDRENS HOSPITAL   
12        74127           NEW HAMPSHIRE HOSP-PSYCH UNIT   
13        61498           UNION COUNTY GENERAL HOSPITAL   
14        59519             MARIA PARHAM MEDICAL CENTER   
15        39426      CHEROKEE INDIAN HOSPITAL AUTHORITY   
16        26704      ALLEGHANY COUNTY MEMORIAL HOSPITAL   
17        71330               AKRON CHILDREN'S HOSPITAL   
18        30887               JEFFERSON COUNTY HOSPITAL   
19        64340                  HOSPITAL SAN FRANCISCO   

                        Address          City State  ZIP Code     County Name  \
0     STATE ROUTE 264 SOUTH 191        GANADO    AZ     86505          APACHE   
1          600 NORTH 7TH STREET  WEST MEMPHIS    AR     72301      CRITTENDEN   
2               5200 NE 2ND AVE         MIAMI    FL     33137      MIAMI-DADE   
3                4480 51ST ST W     BRADENTON    FL     34210         MANATEE   
4      8800 WEST EMERALD STREET         BOISE    ID     83704             ADA   
5   303 NORTH ALLUMBAUGH STREET         BOISE    ID     83704             ADA   
6                1000 N MAIN ST          ANNA    IL     62906           UNION   
7        220 WEST SECOND STREET      GOODLAND    KS     67735         SHERMAN   
8              123 ANDOVER ROAD     WESTBROOK    ME      4092      CUMBERLAND   
9           POST OFFICE BOX 800     CAMBRIDGE    MD     21613      DORCHESTER   
10               8613 MS HWY 12      ACKERMAN    MS     39735         CHOCTAW   
11          ONE CHILDRENS PLACE   SAINT LOUIS    MO     63110  ST. LOUIS CITY   
12            36 CLINTON STREET       CONCORD    NH      3301       MERRIMACK   
13            300 WILSON STREET       CLAYTON    NM     88415           UNION   
14                    PO BOX 59     HENDERSON    NC     27536           VANCE   
15              CALLER BOX C268      CHEROKEE    NC     28719           SWAIN   
16           617 DOCTORS STREET        SPARTA    NC     28675       ALLEGHANY   
17             1 PERKINS SQUARE         AKRON    OH     44308          SUMMIT   
18    INTERSECTION HYWS 81 & 70       WAURIKA    OK     73573       JEFFERSON   
19             371 DE DIEGO AVE      SAN JUAN    PR       923        SAN JUAN   

      Phone Number              Hospital Type              Hospital Ownership  \
0   (928) 755-4541  Critical Access Hospitals  Voluntary non-profit - Private   
1   (870) 394-4113                Psychiatric                     Proprietary   
2   (305) 751-8626       Acute Care Hospitals  Voluntary non-profit - Private   
3   (941) 792-2222                Psychiatric                     Proprietary   
4   (208) 373-5000       Acute Care Hospitals                     Proprietary   
5   (208) 377-8400                Psychiatric                     Proprietary   
6   (618) 833-5161                Psychiatric              Government - State   
7   (785) 890-3625  Critical Access Hospitals              Government - Local   
8   (207) 761-2200                Psychiatric            Government - Federal   
9   (410) 221-2525                Psychiatric              Government - State   
10  (662) 285-4400  Critical Access Hospitals              Government - Local   
11  (314) 454-6000                  Childrens    Voluntary non-profit - Other   
12  (603)

**Note**: If you knew that `hospital_reimbursements` has unique records before conducting an outter join, you could save yourself some headache of dropping by conducting an inner join.

In [17]:
# The DataFrames are switched here because inner join perserves the keys of the left DataFrame and this allows the inner join to be displayed the same as the previous codeblocks.
inner_merged = hospital_reimbursement.merge(hospital_accounts, left_on = 'Provider Name', right_on='Facility Name',  how='inner')
inner_merged.head(20)

Provider_Num                      Provider Name  \
0         839987   SOUTHEAST ALABAMA MEDICAL CENTER   
1         201752           MIZELL MEMORIAL HOSPITAL   
2         678488                  ST VINCENT'S EAST   
3         785513     DEKALB REGIONAL MEDICAL CENTER   
4         890851      SHELBY BAPTIST MEDICAL CENTER   
5         227327     HELEN KELLER MEMORIAL HOSPITAL   
6         881894                DALE MEDICAL CENTER   
7         716815       BAPTIST MEDICAL CENTER SOUTH   
8         187651      JACKSON HOSPITAL & CLINIC INC   
9         895716        EAST ALABAMA MEDICAL CENTER   
10        466874     UNIVERSITY OF ALABAMA HOSPITAL   
11        145182             COMMUNITY HOSPITAL INC   
12        635249    CULLMAN REGIONAL MEDICAL CENTER   
13        693188     STRINGFELLOW MEMORIAL HOSPITAL   
14        371136                HUNTSVILLE HOSPITAL   
15        642438    GADSDEN REGIONAL MEDICAL CENTER   
16        587468     MARION REGIONAL MEDICAL CENTER   
17        538749             FAYETTE MEDICAL CENTER   
18        794509  RIVERVIEW REGIONAL MEDICAL CENTER   
19        794509  RIVERVIEW REGIONAL MEDICAL CENTER   

         Provider Street Address Provider City Provider State  \
0         1108 ROSS CLARK CIRCLE        DOTHAN             AL   
1                  702 N MAIN ST           OPP             AL   
2     50 MEDICAL PARK EAST DRIVE    BIRMINGHAM             AL   
3           200 MED CENTER DRIVE    FORT PAYNE             AL   
4        1000 FIRST STREET NORTH     ALABASTER             AL   
5   1300 SOUTH MONTGOMERY AVENUE     SHEFFIELD             AL   
6               126 HOSPITAL AVE         OZARK             AL   
7      2105 EAST SOUTH BOULEVARD    MONTGOMERY             AL   
8               1725 PINE STREET    MONTGOMERY             AL   
9         2000 PEPPERELL PARKWAY       OPELIKA             AL   
10         619 SOUTH 19TH STREET    BIRMINGHAM             AL   
11           805 FRIENDSHIP ROAD     TALLASSEE             AL   
12      1912 ALABAMA HIGHWAY 157       CULLMAN             AL   
13              301 EAST 18TH ST      ANNISTON             AL   
14                 101 SIVLEY RD    HUNTSVILLE             AL   
15          1007 GOODYEAR AVENUE       GADSDEN             AL   
16    1256 MILITARY STREET SOUTH      HAMILTON             AL   
17      1653 TEMPLE AVENUE NORTH       FAYETTE             AL   
18        600 SOUTH THIRD STREET       GADSDEN             AL   
19        600 SOUTH THIRD STREET       GADSDEN             AL   

    Provider Zip Code  Total Discharges  Average Covered Charges  \
0               36301               118                 20855.61   
1               36467                12                 10901.33   
2               35235                74                 28117.95   
3               35968                19                 28633.79   
4               35007                95                 33169.57   
5               35660                92                 19698.71   
6               36360                15                 10025.87   
7               36116               168                 20290.68   
8               36106               119                 25884.71   
9               36801               164                 12412.13   
10              35233               152                 24991.62   
11              36078                26                  8054.27   
12              35058               107                 19352.06   
13              36201                26                 31828.92   
14              35801               351                 29798.56   
15              35903                74                 73567.61   
16              35570                11                 13623.64   
17              35555                15                 15050.27   
18              35901                40                 36013.58   
19              35901                40                 36013.58   

    Average Total Payments  Average Medicare Payments

Removing the incomplete rows is good, but there's still a discrepancy between the number of overlapping rows and the number of rows in the original `hospital_reimbursement` data.


In [18]:
print("Original Count: {x}".format(x=hospital_reimbursement['Provider Name'].count()))
print("Outter Join Count: {x}".format(x=merged['Provider Name'].count()))
print("Inner Join Count: {x}".format(x=inner_merged['Provider Name'].count()))
print("\nUnaccounted For Rows: {x}".format(x=abs(hospital_reimbursement['Provider Name'].count()-merged['Provider Name'].count())))

Original Count: 2697
Outter Join Count: 2739
Inner Join Count: 2739

Unaccounted For Rows: 42


As we have removed unique rows from our merged data, the remaining unaccounted for rows may be:
- Duplicates caused by two entires having overlapping key names
- Actual hospital accounts and practices sharing the same name

**Optional Challenge**: Try isolating the duplicate entires to see this for yourself.

In [24]:
# Find duplicate entries in the merged DataFrame based on 'Provider Name'
duplicates_in_merged = merged[merged['Provider Name'].duplicated(keep=False)]

print("Duplicate entries in the merged DataFrame:")
display(duplicates_in_merged.sort_values(by='Provider Name'))
print("\nNumber of duplicate entries in the merged DataFrame: {x}".format(x=duplicates_in_merged.shape[0]))

Duplicate entries in the merged DataFrame:


Account_Num            Facility Name                        Address  \
2709      45360.0    AURORA MEDICAL CENTER       975 PORT WASHINGTON ROAD   
2708      45360.0    AURORA MEDICAL CENTER       975 PORT WASHINGTON ROAD   
2340      53660.0  CITIZENS MEDICAL CENTER         7939 U S HWY 165 SOUTH   
2339      30459.0  CITIZENS MEDICAL CENTER            2701 HOSPITAL DRIVE   
2338      71053.0  CITIZENS MEDICAL CENTER            100 E COLLEGE DRIVE   
...           ...                      ...                            ...   
1825      46755.0    WILSON MEDICAL CENTER  REECE CAMPUS 2600 OTTAWA ROAD   
1185      48737.0            YORK HOSPITAL       1001 SOUTH GEORGE STREET   
1186      51583.0            YORK HOSPITAL              15 HOSPITAL DRIVE   
2092      48737.0            YORK HOSPITAL       1001 SOUTH GEORGE STREET   
2093      51583.0            YORK HOSPITAL              15 HOSPITAL DRIVE   

          City State  ZIP Code County Name    Phone Number  \
2709   GRAFTON    WI   53024.0     OZAUKEE  (262) 329-1000   
2708   GRAFTON    WI   53024.0     OZAUKEE  (262) 329-1000   
2340  COLUMBIA    LA   71418.0    CALDWELL  (318) 649-6106   
2339  VICTORIA    TX   77901.0    VICTORIA  (361) 573-9181   
2338     COLBY    KS   67701.0      THOMAS  (785) 462-7511   
...        ...   ...       ...         ...             ...   
1825  NEODESHA    KS   66757.0      WILSON  (620) 325-2611   
1185      YORK    PA   17403.0        YORK  (717) 851-2345   
1186      YORK    ME    3909.0        YORK  (207) 363-4321   
2092      YORK    PA   17403.0        YORK  (717) 851-2345   
2093      YORK    ME    3909.0        YORK  (207) 363-4321   

                  Hospital Type                           Hospital Ownership  \
2709       Acute Care Hospitals               Voluntary non-profit - Private   
2708       Acute Care Hospitals               Voluntary non-profit - Private   
2340       Acute Care Hospitals  Government - Hospital District or Authority   
2339       Acute Care Hospitals                 Voluntary non-profit - Other   
2338  Critical Access Hospitals               Voluntary non-profit - Private   
...                         ...                                          ...   
1825  Critical Access Hospitals                           Government - Local   
1185       Acute Care Hospitals               Voluntary non-profit - Private   
1186       Acute Care Hospitals               Voluntary non-profit - Private   
2092       Acute Care Hospitals               Voluntary non-profit - Private   
2093       Acute Care Hospitals               Voluntary non-profit - Private   

      Provider_Num            Provider Name   Provider Street Address  \
2709        544919    AURORA MEDICAL CENTER  975 PORT WASHINGTON ROAD   
2708        384586    AURORA MEDICAL CENTER        36500 AURORA DRIVE   
2340        199595  CITIZENS MEDICAL CENTER       2701 HOSPITAL DRIVE   
2339        199595  CITIZENS MEDICAL CENTER       2701 HOSPITAL DRIVE   
2338        199595  CITIZENS MEDICAL CENTER       2701 HOSPITAL DRIVE   
...            ...                      ...                       ...   
1825        350387    WILSON MEDICAL CENTER         1705 S TARBORO ST   
1185        585848            YORK HOSPITAL         15 HOSPITAL DRIVE   
1186        585848            YORK HOSPITAL         15 HOSPITAL DRIVE   
2092        888799            YORK HOSPITAL  1001 SOUTH GEORGE STREET   
2093        888799            YORK HOSPITAL  1001 SOUTH GEORGE STREET   

     Provider City Provider State  Provider Zip Code  Total Discharges  \
2709       GRAFTON             WI              53024                46   
2708        SUMMIT             WI              53066                23   
2340      VICTORIA             TX              77901                59   
2339      VICTORIA             TX              77901                59   
2338      VICTORIA             TX              77901                59   
...            ...            ...                ...  


Number of duplicate entries in the merged DataFrame: 616


So how many records have we successfully matched?

In [25]:
merged[merged['Provider Name'].notnull()]['Provider Name'].count()

np.int64(2739)

In the case where the keys in each DataFrame are nicely formatted (no spelling mistakes) performing a direct join like this is possible, but as we've discussed in class this is not always the case.

## Record Linkage
Lets look at the Python [Record Linkage](https://recordlinkage.readthedocs.io/en/latest/) toolkit.



In [26]:
import recordlinkage

We need to create the search space on which records are compared against each other. In class we discussed doing a full comparison (complete pairwise), but also using Blocking and Sorted Neightbourhoods to reduce the complexity. Lets look at how to do all three using the `recordlinkage.Indexer()` (information found [here](https://recordlinkage.readthedocs.io/en/latest/ref-index.html)).

For the `left_on` and `right_on` in `block()` and `sortedneighbourhood()`, we want to use a column to block our data an and to sort our data on, which will create smaller groups in the blocking case and sort the data nicely in the sorted case. Using the "State" column in the data we can separate out the data nicely and it allows us to sort the dataset easily as well.

In [27]:
# Build an indexer to index the datset using different blocking methods
## Full Pairwise comparison
full_indexer = recordlinkage.Index()
full_indexer.full()
## Blocking
block_indexer = recordlinkage.Index()
block_indexer.block(left_on='State', right_on='Provider State')
## Sorted Neighbourhoods
sorted_neigh_index = recordlinkage.Index()
sorted_neigh_index.sortedneighbourhood(left_on='State', right_on='Provider State')

<Index>

In [28]:
full_candidates = full_indexer.index(hospital_accounts, hospital_reimbursement)
print("Full Comparison:  {x}".format(x=len(full_candidates)))
block_candidates = block_indexer.index(hospital_accounts, hospital_reimbursement)
print("Block Comparison:  {x}".format(x=len(block_candidates)))
sorted_neigh_candidates = sorted_neigh_index.index(hospital_accounts, hospital_reimbursement)
print("Sorted Neighbourhood Comparison:  {x}".format(x=len(sorted_neigh_candidates)))

Full Comparison:  14399283
Block Comparison:  475830
Sorted Neighbourhood Comparison:  998860


Looking at our different space reduction techniques we can see that performing a full comparison would result in n * m (5339 * 2697) comparisons whereas blocking and sorted neighbourhoods provide a subset of that search space.

**Note**: If there are a large amount of spelling mistakes use [Sorted Nieghbourhood](https://recordlinkage.readthedocs.io/en/latest/ref-index.html#recordlinkage.index.SortedNeighbourhood) as blocking will fail as it excludes too many records on minor spelling mistakes.





As there aren't a large amount of spelling mistakes (from a cursory glance at the data at least) and blocking results in the least number of comparisons, we will use blocking.

Using the [compare](https://recordlinkage.readthedocs.io/en/latest/ref-compare.html#module-recordlinkage.compare) method we can check if the "City" matches the "Provider City" exactly and if the "Facility Name" matches the "Provider Name", and the "Address" matches the "Prodiver Street Address" using Jaro's similarity set to accept if there is a greater than 85% similarity in both attributes.

**Note**: RecordLinkage has several other string comparison methods implemented, including `jarowinkler` (modification of Jaro similarity), `levenshtein` (edit distance), `damerau_levenshtein` (modification of edit distance), `qgram` (ngram), and `cosine`. The threshold is used for similarity, so for the distance functions (i.e. edit, n-gram) it's 1-distance.

In [29]:
compare = recordlinkage.Compare()
compare.exact('City', 'Provider City', label='City')
compare.string('Facility Name',
               'Provider Name',
               method='jaro',
               threshold=0.85,
               label='Hosp_Name')
compare.string('Address',
               'Provider Street Address',
               method='jaro',
               threshold=0.85,
               label='Hosp_Address')
features = compare.compute(block_candidates, hospital_accounts,
                           hospital_reimbursement)

In [30]:
features.sort_values(["City", "Hosp_Name", "Hosp_Address"], ascending=False)

City  Hosp_Name  Hosp_Address
14   1777     1        1.0           1.0
27   1502     1        1.0           1.0
28   779      1        1.0           1.0
34   2170     1        1.0           1.0
38   1308     1        1.0           1.0
...         ...        ...           ...
5338 1435     0        0.0           0.0
     1436     0        0.0           0.0
     1437     0        0.0           0.0
     1438     0        0.0           0.0
     1439     0        0.0           0.0

[475830 rows x 3 columns]

The compare method returns a 1 if the similarity is greater than the threshold and a 0 otherwise. If all three comparisons were 85% similar, then we can be fairly sure the two records represent the same real-world entity.

The two columns to the left without a title are the indices of these records in the orignal table. By using the reset_index() method we can move these columns into the DataFrame, and manually reset the column title to reflect what it represents.

In [31]:
# Get the potential matches
potential_matches = features[features.sum(axis=1) >= 3].reset_index()
potential_matches.rename(columns={"level_0":"account_index", "level_1": "reimbursement_index"}, inplace=True)
potential_matches.head(20)

account_index  reimbursement_index  City  Hosp_Name  Hosp_Address
0              14                 1777     1        1.0           1.0
1              27                 1502     1        1.0           1.0
2              28                  779     1        1.0           1.0
3              34                 2170     1        1.0           1.0
4              38                 1308     1        1.0           1.0
5              40                 2407     1        1.0           1.0
6              42                 2063     1        1.0           1.0
7              43                  997     1        1.0           1.0
8              44                  794     1        1.0           1.0
9              47                 2044     1        1.0           1.0
10             52                 1367     1        1.0           1.0
11             55                  854     1        1.0           1.0
12             56                  206     1        1.0           1.0
13             59                  826     1        1.0           1.0
14             61                   83     1        1.0           1.0
15             66                 1216     1        1.0           1.0
16             67                 1228     1        1.0           1.0
17             69                 1534     1        1.0           1.0
18             72                 1723     1        1.0           1.0
19             74                 2106     1        1.0           1.0

Now, lets have a look at a couple records and see if they seem to represent the same thing.

In [32]:
hospital_accounts.loc[14,:]

Account_Num                                 59519
Facility Name         MARIA PARHAM MEDICAL CENTER
Address                                 PO BOX 59
City                                    HENDERSON
State                                          NC
ZIP Code                                    27536
County Name                                 VANCE
Phone Number                       (252) 431-3708
Hospital Type                Acute Care Hospitals
Hospital Ownership                    Proprietary
Name: 14, dtype: object

In [33]:
hospital_reimbursement.loc[1777,:]

Provider_Num                                      603343
Provider Name                MARIA PARHAM MEDICAL CENTER
Provider Street Address                        PO BOX 59
Provider City                                  HENDERSON
Provider State                                        NC
Provider Zip Code                                  27536
Total Discharges                                      70
Average Covered Charges                         18270.17
Average Total Payments                            6180.2
Average Medicare Payments                        4849.11
Name: 1777, dtype: object

Record 14 in `hospital_accounts` and `hospital_reimbursement` do both appear to represent the same real-world hospital, lets check out another set of records to see if that's also the case.

In [34]:
hospital_accounts.loc[27,:]

Account_Num                                       71617
Facility Name         NEWARK BETH ISRAEL MEDICAL CENTER
Address                                   201 LYONS AVE
City                                             NEWARK
State                                                NJ
ZIP Code                                           7112
County Name                                       ESSEX
Phone Number                             (973) 926-7850
Hospital Type                      Acute Care Hospitals
Hospital Ownership       Voluntary non-profit - Private
Name: 27, dtype: object

In [35]:
hospital_reimbursement.loc[1502,:]

Provider_Num                                            123841
Provider Name                NEWARK BETH ISRAEL MEDICAL CENTER
Provider Street Address                          201 LYONS AVE
Provider City                                           NEWARK
Provider State                                              NJ
Provider Zip Code                                         7112
Total Discharges                                           146
Average Covered Charges                               70661.73
Average Total Payments                                11955.77
Average Medicare Payments                             10009.51
Name: 1502, dtype: object

Same thing with the second record. If we continue through all fo the records with an aggregate similarity score of 3, we will see that they all seem to correspond to matching entities.

Now lets see what the total number of each score was across all of the comparisons made.

In [36]:
# What are the score totals?
features.sum(axis=1).value_counts().sort_index(ascending=False)

3.0      2299
2.0       489
1.0      8530
0.0    464512
Name: count, dtype: int64

RecordLinkage was able to perfectly match 2299 records, with 489 strong partial matches (should be manually reviewed) 8530 weak partial matches (should be manually reviewed by an intern), and 464512 non matches.

Now that we've come to the end of our RecordLinkage section, I hope you've enjoyed the differences between manually joining tables and performing record linkage.

**Optional Challenge**: If you want an extra challenge, try playing around with the match threshold levels to see how the partial matches change, as well as using other similarity metrics.

## Fuzzymatcher
Just as another example of a record linking tool here is fuzzymatcher. It is unfortunately no longer being supported but I've left this short example in as it was one of the examples used in last year's practical.

In [37]:
import fuzzymatcher
# Columns to match on from the hospital_accounts df
left_on = ["Facility Name", "Address", "City", "State"]
# Columns to match on from the hostpital_reimbursements df
right_on = ["Provider Name", "Provider Street Address", "Provider City", "Provider State"]
# Try link_table first - this may take a while have
linked = fuzzymatcher.link_table(hospital_accounts,
                                 hospital_reimbursement,
                                 left_on,
                                 right_on,
                                 left_id_col='Account_Num',
                                 right_id_col='Provider_Num')
linked.head()

__id_left  __id_right  match_score  match_rank           Facility Name  \
0      10605      643595    -0.746613           1  SAGE MEMORIAL HOSPITAL   
1      10605      766793    -0.820139           2  SAGE MEMORIAL HOSPITAL   
2      10605      572806    -0.853538           3  SAGE MEMORIAL HOSPITAL   
3      10605      489749    -0.877983           4  SAGE MEMORIAL HOSPITAL   
4      10605      264631    -0.890167           5  SAGE MEMORIAL HOSPITAL   

                Provider Name                    Address  \
0     TYLER MEMORIAL HOSPITAL  STATE ROUTE 264 SOUTH 191   
1           ANDERSON HOSPITAL  STATE ROUTE 264 SOUTH 191   
2     MERCY HOSPITAL ANDERSON  STATE ROUTE 264 SOUTH 191   
3  MONTROSE MEMORIAL HOSPITAL  STATE ROUTE 264 SOUTH 191   
4     JONES MEMORIAL HOSPITAL  STATE ROUTE 264 SOUTH 191   

   Provider Street Address    City Provider City State Provider State  
0  5950 STATE ROUTE 6 WEST  GANADO   TUNKHANNOCK    AZ             PA  
1     6800 STATE ROUTE 162  GANADO     MARYVILLE    AZ             IL  
2          7500 STATE ROAD  GANADO    CINCINNATI    AZ             OH  
3             800 S 3RD ST  GANADO      MONTROSE    AZ             CO  
4    191 NORTH MAIN STREET  GANADO    WELLSVILLE    AZ             NY

In [38]:
linked[linked['Facility Name'].str.contains('GORDON')].sort_values(by='match_score', ascending=False)

__id_left  __id_right  match_score  match_rank  \
310640      24125      515652     0.986440           1   
309991      24261      693188    -0.254650           1   
309992      24261      421579    -0.573855           2   
309993      24261      340833    -0.615630           3   
309994      24261      522917    -0.654419           4   
...           ...         ...          ...         ...   
310202      24261      875973    -1.751271         212   
310203      24261      236029    -1.803042         213   
310204      24261      683090    -1.820631         214   
310205      24261      198667    -1.922763         215   
310206      24261      610170    -1.981611         216   

                            Facility Name  \
310640                ADVENTHEALTH GORDON   
309991  GORDON MEMORIAL HOSPITAL DISTRICT   
309992  GORDON MEMORIAL HOSPITAL DISTRICT   
309993  GORDON MEMORIAL HOSPITAL DISTRICT   
309994  GORDON MEMORIAL HOSPITAL DISTRICT   
...                                   ...   
310202  GORDON MEMORIAL HOSPITAL DISTRICT   
310203  GORDON MEMORIAL HOSPITAL DISTRICT   
310204  GORDON MEMORIAL HOSPITAL DISTRICT   
310205  GORDON MEMORIAL HOSPITAL DISTRICT   
310206  GORDON MEMORIAL HOSPITAL DISTRICT   

                                    Provider Name            Address  \
310640                            GORDON HOSPITAL  1035 RED BUD ROAD   
309991             STRINGFELLOW MEMORIAL HOSPITAL    300 EAST 8TH ST   
309992                COLUMBUS COMMUNITY HOSPITAL    300 EAST 8TH ST   
309993                  RIDEOUT MEMORIAL HOSPITAL    300 EAST 8TH ST   
309994              BAKERSFIELD MEMORIAL HOSPITAL    300 EAST 8TH ST   
...                                           ...                ...   
310202            ST ELIZABETH COMMUNITY HOSPITAL    300 EAST 8TH ST   
310203       PEACEHEALTH SOUTHWEST MEDICAL CENTER    300 EAST 8TH ST   
310204                       CLAY COUNTY HOSPITAL    300 EAST 8TH ST   
310205                    LOVELACE MEDICAL CENTER    300 EAST 8TH ST   
310206  CHINLE COMPREHENSIVE HEALTH CARE FACILITY    300 EAST 8TH ST   

                    Provider Street Address     City Provider City State  \
310640                    1035 RED BUD ROAD  CALHOUN       CALHOUN    GA   
309991                     301 EAST 18TH ST   GORDON      ANNISTON    NE   
309992                         4600 38TH ST   GORDON      COLUMBUS    NE   
309993                           726 4TH ST   GORDON    MARYSVILLE    NE   
309994                      420 34TH ST BOX   GORDON   BAKERSFIELD    NE   
...                                     ...      ...           ...   ...   
310202       2550 SISTER MARY COLUMBA DRIVE   GORDON     RED BLUFF    NE   
310203           400 NE MOTHER JOSEPH PLACE   GORDON     VANCOUVER    NE   
310204      83825 HIGHWAY 9    P O BOX 1270   GORDON       ASHLAND    NE   
310205  601 DR MARTIN LUTHER KING JR AVE NE   GORDON   ALBUQUERQUE    NE   
310206            US HWY 191, HOSPITAL ROAD   GORDON        CHINLE    NE   

       Provider State  
310640             GA  
309991             AL  
309992             NE  
309993             CA  
309994             CA  
...               ...  
310202             CA  
310203             WA  
310204             AL  
310205             NM  
310206             AZ  

[217 rows x 12 columns]

We can see that the `match_score` column shows how certain the FuzzyMatcher is that two records represent the same thing. Looking at the first record, we can see that the address from both files is the same, and while the "Facility Name" and "Provider Name" are slightly different, they probably represent the same real-world entity. We wouldn't have been able to make this match just using a single join, so looking across multiple columns like we have here can make better Record Linkage decisions.

Now lets use a `fuzzy_left_join` to combine our two datasets together and see the best and worst matches bewteen the two sets.

In [39]:
# It will take several minutes to run on this data set
matched_results = fuzzymatcher.fuzzy_left_join(hospital_accounts,
                                               hospital_reimbursement,
                                               left_on,
                                               right_on,
                                               left_id_col='Account_Num',
                                               right_id_col='Provider_Num')

In [40]:
# Reorder the columns to make viewing easier
cols = [
    "best_match_score", "Facility Name", "Provider Name", "Address", "Provider Street Address",
    "Provider City", "City", "Provider State", "State"
]

In [41]:
# Let's see the best 5 matches
matched_results[cols].sort_values(by=['best_match_score'], ascending=False).head(5)

best_match_score                                      Facility Name  \
78296           3.090931    RARITAN BAY MEDICAL CENTER PERTH AMBOY DIVISION   
533136          2.799072            ROBERT WOOD JOHNSON UNIVERSITY HOSPITAL   
78906           2.785132  AVERA MCKENNAN HOSPITAL & UNIVERSITY HEALTH CE...   
242735          2.778860  JOHN T MATHER MEMORIAL HOSPITAL  OF PORT JEFFE...   
447854          2.721425               MAYO CLINIC HEALTH SYSTEM - RED WING   

                                            Provider Name  \
78296     RARITAN BAY MEDICAL CENTER PERTH AMBOY DIVISION   
533136            ROBERT WOOD JOHNSON UNIVERSITY HOSPITAL   
78906   AVERA MCKENNAN HOSPITAL & UNIVERSITY HEALTH CE...   
242735  JOHN T MATHER MEMORIAL HOSPITAL  OF PORT JEFFE...   
447854              MAYO CLINIC HEALTH SYSTEM IN RED WING   

                                       Address  \
78296                    530 NEW BRUNSWICK AVE   
533136           ONE ROBERT WOOD JOHNSON PLACE   
78906   1325 S CLIFF AVE  POST OFFICE BOX 5045   
242735                   75 NORTH COUNTRY ROAD   
447854         701 HEWITT BOULEVARD, PO BOX 95   

                       Provider Street Address   Provider City  \
78296                    530 NEW BRUNSWICK AVE     PERTH AMBOY   
533136           ONE ROBERT WOOD JOHNSON PLACE   NEW BRUNSWICK   
78906   1325 S CLIFF AVE  POST OFFICE BOX 5045     SIOUX FALLS   
242735                   75 NORTH COUNTRY ROAD  PORT JEFFERSON   
447854         701 HEWITT BOULEVARD, PO BOX 95        RED WING   

                  City Provider State State  
78296      PERTH AMBOY             NJ    NJ  
533136   NEW BRUNSWICK             NJ    NJ  
78906      SIOUX FALLS             SD    SD  
242735  PORT JEFFERSON             NY    NY  
447854        RED WING             MN    MN

In [42]:
# Now the worst matches
matched_results[cols].sort_values(by=['best_match_score'], ascending=True).head(5)

best_match_score                                      Facility Name  \
426629         -2.268231                      CENTRO MEDICO WILMA N VAZQUEZ   
83501          -2.124071  DOCTOR CENTER HOSPITAL SAN FERNANDO DE LA CARO...   
42861          -2.106746     HOSPITAL ONCOLOGICO DR ISAAC GONZALEZ MARTINEZ   
450732         -2.050888           CENTRO DE SALUD CONDUCTUAL MENONITA-CIMA   
476484         -1.996508             ADMIN DE SERVICIOS MEDICOS  PUERTO RIC   

                           Provider Name  \
426629       BAPTIST MEDICAL CENTER EAST   
83501   OVERLAKE HOSPITAL MEDICAL CENTER   
42861             SCRIPPS MERCY HOSPITAL   
450732   MILFORD REGIONAL MEDICAL CENTER   
476484              MAINE MEDICAL CENTER   

                                                  Address  \
426629      CARR. 2 KM 39.5    ROAD NUMBER 2 BO ALGARROBO   
83501   EDIF JESUS T PINEIRO AVE FERNANDEZ JUNCOS BO P...   
42861   BO. MONACILLOS CARR 22 CENTRO MEDICO DE PUERTO...   
450732  CARR ESTATAL 14 INTERIOR SARGENTO GERARDO SANT...   
476484                           BO MONACILLO CARR NUM 22   

       Provider Street Address Provider City       City Provider State State  
426629         400 TAYLOR ROAD    MONTGOMERY  VEGA BAJA             AL    PR  
83501        1035-116TH AVE NE      BELLEVUE   CAROLINA             WA    PR  
42861             4077 5TH AVE     SAN DIEGO   SAN JUAN             CA    PR  
450732      14 PROSPECT STREET       MILFORD   AIBONITO             MA    PR  
476484          22 BRAMHALL ST      PORTLAND   SAN JUAN             ME    PR

There is a tool called [Splink](https://moj-analytical-services.github.io/splink/index.html) which is being actively contributed to while FuzzyMatcher is no longer supported. I did not have enough time to create a brief splink tutorial, and the one from last year is depreciated and no longer functions, but if you have the inclination please give it a try in your spare time.